### ISSUE TITLE
MMA alignment_train_cuda_binding does not compute gradient for p_choose tensor

### ISSUE THREAD

**Issue by George0828Zhang** [facebookresearch/fairseq#4270]
## 🐛 Bug

I noticed that the `alignment_train_cuda_binding` does not compute gradients. Specifically, when inspecting backward gradient at [here](https://github.com/pytorch/fairseq/blob/d03f4e771484a433f025f47744017c2eb6e9c6bc/examples/simultaneous_translation/utils/monotonic_attention.py#L52) using pdb, you can see that p_choose does not receive any gradient...

Any model that uses hard attention is **not learning useful alignments**. For instance, MMA-H failed to train at all, while MMA-IL can be trained because it has soft attention, but the **latency cannot be optimized**. You can confirm this by comparing training curve with or without latency training, where they appear identical.

Reverting to earlier commit fixes this issue, but at the cost of 2x slower training.

Judging from the lack of backward functions in the cpp codes for `alignment_train_cuda_binding`, and that the testing code only tests forward pass, it does seem like the author(s) forgot to implement backward pass.

It's really frustrating to have to waste **hundreds** of GPU hours to find out about this :weary:. I hope anyone watching won't have to waste their time (and money) like I did.

Besides, there're a handful of bugs in inference  #4247 and during build #4085. The MMA has been broken for a while.

### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. train any MMA-H models, you can see that the model barely learns anything.
2. train any models with the latency training with any latency weights. The resulting training curve is exactly the same as that without latency training.

### Environment

 - fairseq Version (e.g., 1.0 or main):
 - PyTorch Version (e.g., 1.0)
 - OS (e.g., Linux):
 - How you installed fairseq (`pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by George0828Zhang** [comment]
@xutaima @kahne @renganxu, can you show some support please? Thank you.

**Comment by EricLina** [comment]
I've been reproducing MMA for two months ... there are indeed a lot of issues to be handed

**Comment by EricLina** [comment]
I followed your PR  @George0828Zhang  and MMA-wait looks OK , But MMA-hard is as bad as you said . Hoping for author's
 reply

**Comment by EricLina** [comment]
I also noticed that when trianing MMA-hard ,there attended to be  gradient overflow 
> NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 32.0
